upx-recovery-tool.py

```diff
@@ -0,0 +1,340 @@
+#! /usr/bin/python3
+
+import os
+import re
+import mmap
+import magic
+import shutil
+import struct
+import argparse
+from elftools.elf.elffile import ELFFile
+
+
+class UnsupportedFileError(Exception):
+    pass
+
+
+class NonUpxError(Exception):
+    pass
+
+
+class CorruptedFileError(Exception):
+    pass
+
+
+class l_info_s:
+    l_checksum: bytes # checksum
+    l_magic: bytes # UPX! magic[55 50 58 21]
+    l_lsize: bytes # loader size
+    l_version: bytes # version info
+    l_format: bytes # UPX format
+
+    def __init__(self, buff):
+        self.l_checksum = buff[0:4]
+        self.l_magic = buff[4:8]
+        self.l_lsize = buff[8:10]
+        self.l_version = buff[10:11]
+        self.l_version = buff[11:12]
+
+
+class p_info_s:
+    p_progid: bytes
+    p_filesize: bytes # Size of the unpacked file
+    p_blocksize: bytes # Size of the unpacked file
+
+    def __init__(self, buff):
+        self.p_progid = buff[0:4]
+        self.p_filesize = buff[4:8]
+        self.p_blocksize = buff[8:12]
+        
+
+class PackHeader:
+    u_file_size: bytes
+
+    def __init__(self, file_mmap):
+        ph_offset = file_mmap.rfind(b"UPX!")
+
+        if ph_offset > file_mmap.size() - 36:
+            raise CorruptedFileError("Parsing PackHeader")
+
+        elif ph_offset < file_mmap.size() - 36:
```

Comment:
Just to confirm, the only valid option would be `ph_offset == file_mmap.size() - 36`, right?